presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayIntersectFunction.java

```diff
@@ -0,0 +1,169 @@
+    }
+
+    private static IntComparator IntBlockCompare(Type type, Block inBlock)
+    {
+        return new IntComparator() {
+            @Override
+            public int compare(int l, int r)
+            {
+                if (inBlock.isNull(l) && inBlock.isNull(r)) {
+                    return 0;
+                }
+                if (inBlock.isNull(l)) {
+                    return -1;
+                }
+                if (inBlock.isNull(r)) {
+                    return 1;
+                }
+                return type.compareTo(inBlock, l, inBlock, r);
+            }
+
+            @Override
+            public int compare(Integer l, Integer r)
+            {
+                if (inBlock.isNull(l) && inBlock.isNull(r)) {
+                    return 0;
+                }
+                if (inBlock.isNull(l)) {
+                    return -1;
+                }
+                if (inBlock.isNull(r)) {
+                    return 1;
+                }
+                return type.compareTo(inBlock, l, inBlock, r);
+            }
+        };
+    }
+
+    public static Slice intersect(Type type, Slice leftArray, Slice rightArray)
+    {
+        Block leftBlock = readStructuralBlock(leftArray);
+        Block rightBlock = readStructuralBlock(rightArray);
+
+        int leftPositionCount = leftBlock.getPositionCount();
+        int rightPositionCount = rightBlock.getPositionCount();
+
+        int[] leftPositions = new int[leftPositionCount];
+        int[] rightPositions = new int[rightPositionCount];
+
+        for (int i = 0; i < leftPositionCount; i++) {
+            leftPositions[i] = i;
+        }
+        for (int i = 0; i < rightPositionCount; i++) {
+            rightPositions[i] = i;
+        }
+        IntArrays.quickSort(leftPositions, IntBlockCompare(type, leftBlock));
+        IntArrays.quickSort(rightPositions, IntBlockCompare(type, rightBlock));
+
+        BlockBuilder resultBlockBuilder = new VariableWidthBlockBuilder(new BlockBuilderStatus(), leftBlock.getSizeInBytes());
+
+        int leftCurPos = 0;
+        int rightCurPos = 0;
+
+        while (leftCurPos < leftPositionCount && leftBlock.isNull(leftPositions[leftCurPos])) {
+            leftCurPos++;
+        }
+        while (rightCurPos < rightPositionCount && rightBlock.isNull(rightPositions[rightCurPos])) {
+            rightCurPos++;
+        }
+        while (leftCurPos < leftPositionCount && rightCurPos < rightPositionCount) {
+            int compareValue = type.compareTo(leftBlock, leftPositions[leftCurPos], rightBlock, rightPositions[rightCurPos]);
+            if (compareValue > 0) {
+                rightCurPos++;
+            }
+            else if (compareValue < 0) {
+                leftCurPos++;
+            }
+            else {
+                type.appendTo(leftBlock, leftPositions[leftCurPos], resultBlockBuilder);
+                leftCurPos++;
```

Comment:
You might need to advance more than one position here. Otherwise you might output a duplicate. For example: `array_intersect(array[1, 1], array[1, 1])` should be `array[1]` not `array[1, 1]`